features/environment.py

```diff
@@ -394,92 +470,96 @@ def create_uat_lxd_image(context: Context, series: str) -> None:
+    else:
+        is_vm = bool(context.config.machine_type == "lxd.vm")
+        build_container_name = "behave-image-build-%s-%s" % (
+            "-vm" if is_vm else "",
+            series + time_suffix,
+        )
+        if is_vm and series == "xenial":
+            # FIXME: use lxd custom cloud images which containt HWE kernel for
+            # vhost-vsock support
+            lxc_ubuntu_series = "images:ubuntu/16.04/cloud"
+        else:
+            lxc_ubuntu_series = "ubuntu-daily:%s" % series
         launch_lxd_container(
             context,
             series=series,
-            image_name=ubuntu_series,
+            image_name=lxc_ubuntu_series,
             container_name=build_container_name,
             is_vm=is_vm,
         )
-        lxc_build_deb(build_container_name, output_deb_file=deb_file)
 
-    build_container_name = "behave-image-build%s-%s" % (
-        "-vm" if is_vm else "",
-        series + now.strftime("%s%f"),
+    _install_uat_in_container(
+        build_container_name, debs=debs, cloud_api=context.config.cloud_api
     )
 
-    launch_lxd_container(
-        context,
-        series=series,
-        image_name=ubuntu_series,
-        container_name=build_container_name,
-        is_vm=is_vm,
-    )
-
-    # if build_pr it will install new built .deb
-    _install_uat_in_container(build_container_name, deb_file=deb_file)
-
-    context.series_image_name[
-        series
-    ] = "behave-image-%s-" % series + now.strftime("%s%f")
-
-    _capture_container_as_image(
-        build_container_name, context.series_image_name[series]
+    image_name = _capture_container_as_image(
+        build_container_name,
+        image_name="behave-image-%s-" % series + time_suffix,
+        cloud_api=context.config.cloud_api,
     )
+    context.series_image_name[series] = image_name
 
 
 def _install_uat_in_container(
-    container_name: str, deb_file: "Optional[str]"
+    container_name: str,
+    debs: "Optional[List[str]]" = [],
+    cloud_api: "Optional[pycloudlib.cloud.BaseCloud]" = None,
 ) -> None:
     """Install ubuntu-advantage-tools into the specified container
 
     :param container_name:
         The name of the container into which ubuntu-advantage-tools should be
         installed.
-    :param deb_file: Optional path to the deb_file we need to install
+    :param debs: Optional paths to local deb files we need to install
+    :param cloud_api: Optional pycloud BaseCloud api for applicable
+        machine_types.
     """
-    if deb_file:
-        subprocess.run(
-            ["lxc", "file", "push", deb_file, container_name + "/tmp/"]
-        )
-        lxc_exec(container_name, ["sudo", "dpkg", "-i", deb_file])
-        lxc_exec(
-            container_name, ["apt-cache", "policy", "ubuntu-advantage-tools"]
-        )
-    else:
-        lxc_exec(
-            container_name,
+    cmds = []
+    if not debs:
+        debs = " ".join(UA_DEBS) if cloud_api else "ubuntu-advantage-tools"
```

Comment:
Maybe I am missing something, but isn't the debs we received here already filtered ?

Reply:
We don't actually filter deb paths return valid from build_debs. The function util.lxc_build_debs now actually builds, copies and returns the path of both ubuntu-advantage-tools and *-pro deb files. So we do need to filter somewhere. But, I think it makes more sense to filter this list before we call _install_uat_in_container so that this function can remain simpler. What do you think about me pulling it up out of this function before the _install_uat_in_container call site?